google-cloud-firestore/src/main/java/com/google/cloud/firestore/UserDataConverter.java

```diff
@@ -227,4 +228,45 @@ static Object decodeValue(FirestoreRpcContext<?> rpcContext, Value v) {
             String.format("Unknown Value Type: %s", typeCase));
     }
   }
+
+  static Object decodeGoogleProtobufValue(com.google.protobuf.Value v) {
+    switch (v.getKindCase()) {
+      case NULL_VALUE:
+        return null;
+      case BOOL_VALUE:
+        return v.getBoolValue();
+      case NUMBER_VALUE:
+        return v.getNumberValue();
+      case STRING_VALUE:
+        return v.getStringValue();
+      case LIST_VALUE:
+        List<Object> list = new ArrayList<>();
+        List<com.google.protobuf.Value> lv = v.getListValue().getValuesList();
+        for (com.google.protobuf.Value iv : lv) {
+          list.add(decodeGoogleProtobufValue(iv));
+        }
+        return list;
+      case STRUCT_VALUE:
+        Map<String, Object> outputMap = new HashMap<>();
+        Map<String, com.google.protobuf.Value> inputMap = v.getStructValue().getFieldsMap();
+        for (Map.Entry<String, com.google.protobuf.Value> entry : inputMap.entrySet()) {
+          outputMap.put(entry.getKey(), decodeGoogleProtobufValue(entry.getValue()));
+        }
+        return outputMap;
```

Comment:
Can I suggest we use Gauva here? Two reasons:
1. Elagent syntax.
2. This is a lazy view of underlying collections. This can avoid converting a large data structure when only small parts are plucked out later.

```suggestion
      case LIST_VALUE:
        return Lists.transform(
            v.getListValue().getValuesList(),
            UserDataConverter::decodeGoogleProtobufValue
        );
      case STRUCT_VALUE:
        return Maps.transformValues(
            v.getStructValue().getFieldsMap(),
            UserDataConverter::decodeGoogleProtobufValue
        );
```